File: src/lib/seriesGrouping.ts
/** Group individual OE games into Bo3/Bo5 series by matchup, chronology, and completion. */

export const SERIES_GAP_DAYS = 5

export interface ChronologicalGame {
  id: string
  date: string
  winner: string
  loser: string
}

export interface SeriesBucket<T extends ChronologicalGame = ChronologicalGame> {
  teamA: string
  teamB: string
  games: T[]
  /** Disambiguates multiple series between the same pair on the same day. */
  sessionIndex: number
}

export function seriesKey(a: string, b: string): string {
  return [a, b].sort((x, y) => x.localeCompare(y)).join('|')
}

export function canonicalSeriesTeams(a: string, b: string): [string, string] {
  return [a, b].sort((x, y) => x.localeCompare(y)) as [string, string]
}

function parseDate(value: string): Date | null {
  const d = new Date(value)
  return Number.isNaN(d.getTime()) ? null : d
}

export function daysBetween(a: string, b: string): number {
  const da = parseDate(a)
  const db = parseDate(b)
  if (!da || !db) return 999
  return Math.abs(da.getTime() - db.getTime()) / (1000 * 60 * 60 * 24)
}

/** Match OE game ordering: date, then game id ordinal, then id string. */
export function gameOrdinalFromId(id: string): number {
  const tail = id.match(/[_-](\d+)$/)
  if (tail) return Number(tail[1])
  const embedded = id.match(/game[_-]?(\d+)/i)
  if (embedded) return Number(embedded[1])
  return 0
}

export function compareSeriesGames(
  a: { date: string; id: string },
  b: { date: string; id: string },
): number {
  const byDate = a.date.localeCompare(b.date)
  if (byDate !== 0) return byDate
  const byOrd = gameOrdinalFromId(a.id) - gameOrdinalFromId(b.id)
  if (byOrd !== 0) return byOrd
  return a.id.localeCompare(b.id)
}

export function countSeriesWins(games: ChronologicalGame[], team: string): number {
  return games.filter((g) => g.winner === team).length
}

/** True when game order respects first-to-2 (Bo3) / first-to-3 (Bo5) progression. */
export function isValidGameOrder(
  games: ChronologicalGame[],
  teamA: string,
  teamB: string,
): boolean {
  let wA = 0
  let wB = 0
  for (let i = 0; i < games.length; i++) {
    const g = games[i]!
    if (g.winner === teamA) wA++
    else if (g.winner === teamB) wB++
    else return false

    if (wA === 3 || wB === 3) {
      return i === games.length - 1
    }
  }

  return isValidSeriesScore(wA, wB)
}

function permuteGames<T>( games: T[], maxPerms = 5040): T[][] {
  if (games.length <= 1) return [games]
  if (games.length > 7) return [games]

  const out: T[][] = []
  const arr = [...games]

  const walk = (k: number) => {
    if (out.length >= maxPerms) return
    if (k === 1) {
      out.push([...arr])
      return
    }
    for (let i = 0; i < k; i++) {
      walk(k - 1)
      if (k % 2 === 0) {
        ;[arr[i], arr[k - 1]] = [arr[k - 1]!, arr[i]!]
      } else {
        ;[arr[0], arr[k - 1]] = [arr[k - 1]!, arr[0]!]
      }
    }
  }

  walk(arr.length)
  return out
}

function orderDistance(a: ChronologicalGame[], b: ChronologicalGame[]): number {
  const indexB = new Map(b.map((g, i) => [g.id, i]))
  let dist = 0
  for (let i = 0; i < a.length; i++) {
    const j = indexB.get(a[i]!.id) ?? i
    dist += Math.abs(i - j)
  }
  return dist
}

/**
 * OE gameid suffixes are not always chronological within a series-day cluster.
 * Find a valid Bo3/Bo5 progression; prefer permutations closest to game-id sort.
 */
export function orderSeriesGames<T extends ChronologicalGame>(
  games: T[],
  teamA: string,
  teamB: string,
): T[] {
  if (games.length <= 1) return games

  const byId = [...games].sort(compareSeriesGames)
  if (isValidGameOrder(byId, teamA, teamB)) return byId

  const valid = permuteGames(byId).filter((order) => isValidGameOrder(order, teamA, teamB))
  if (!valid.length) return byId

  valid.sort((a, b) => orderDistance(a, byId) - orderDistance(b, byId))
  return valid[0]!
}

function clusterByDate<T extends ChronologicalGame>(games: T[]): T[][] {
  const sorted = [...games].sort(compareSeriesGames)
  const clusters: T[][] = []
  let current: T[] = []

  for (const g of sorted) {
    if (!current.length || current[current.length - 1]!.date === g.date) {
      current.push(g)
    } else {
      clusters.push(current)
      current = [g]
    }
  }

  if (current.length) clusters.push(current)
  return clusters
}

/** Valid tier-1 series scores (Bo3 or Bo5). */
export function isValidSeriesScore(wA: number, wB: number): boolean {
  const max = Math.max(wA, wB)
  const min = Math.min(wA, wB)
  if (max >= 3) return max === 3 && min <= 2
  if (max === 2) return min <= 1
  return false
}

/**
 * 2-0 / 2-1 can be a finished Bo3 OR an incomplete Bo5 when OE is mid-series.
 * Prefer Cito schedule confirmation before treating these as final (see citoSeriesVerify).
 */
export function isProvisionalSeriesScore(wA: number, wB: number): boolean {
  const max = Math.max(wA, wB)
  const min = Math.min(wA, wB)
  return max === 2 && min <= 1
}

/**
 * Scores allowed on tournament/series lists: terminal Bo3/Bo5, OR mid-Bo5 shapes
 * (including 2-2) when OE lags behind a live international series.
 */
export function isListableSeriesScore(
  wA: number,
  wB: number,
  opts?: { allowInProgress?: boolean },
): boolean {
  if (isValidSeriesScore(wA, wB)) return true
  if (!opts?.allowInProgress) return false
  const max = Math.max(wA, wB)
  const total = wA + wB
  return total >= 1 && max < 3 && total <= 5
}

/**
 * True when cumulative wins match a completed Bo3 or Bo5.
 * Pass `nextGame` for lookahead — a 2-1 after 3 games is still in progress if another
 * game between the same teams follows within SERIES_GAP_DAYS (Bo5 games 4–5).
 */
export function isSeriesComplete(
  games: ChronologicalGame[],
  teamA: string,
  teamB: string,
  nextGame?: ChronologicalGame,
): boolean {
  const wA = countSeriesWins(games, teamA)
  const wB = countSeriesWins(games, teamB)
  const max = Math.max(wA, wB)
  const min = Math.min(wA, wB)
  const total = wA + wB

  if (max === 3) return true

  if (max !== 2 || !isValidSeriesScore(wA, wB)) return false

  // 2-0 or 2-1 may still be an in-progress Bo5 when more games follow soon.
  const bo5InProgress =
    (total === 2 && min === 0) || (total === 3 && min === 1)

  if (!bo5InProgress) return true

  // Critical: when OE has not yet published the next game, do NOT close the series.
  // Closing early caused game 3 of a Bo5 to start a new 1-0 bucket instead of extending 2-0.
  if (!nextGame) return false

  if (daysBetween(games[games.length - 1]!.date, nextGame.date) > SERIES_GAP_DAYS) {
    return true
  }

  // Cross-day rematch: completed Bo3 (2-1 after 3 games) vs next-day Bo5 continuation.
  // Break when the series sat complete overnight (>=1 day gap) at a Bo3 terminal score.
  const lastDate = games[games.length - 1]!.date
  const dayGap = daysBetween(lastDate, nextGame.date)
  if (dayGap >= 1 && total === 3 && min === 1) {
    return true
  }

  return false
}

export function shouldBreakSeries(
  bucketGames: ChronologicalGame[],
  newGame: ChronologicalGame,
  teamA: string,
  teamB: string,
): boolean {
  if (!bucketGames.length) return false
  const last = bucketGames[bucketGames.length - 1]!
  if (daysBetween(last.date, newGame.date) > SERIES_GAP_DAYS) return true
  return isSeriesComplete(bucketGames, teamA, teamB, newGame)
}

function splitPairGamesIntoSeries<T extends ChronologicalGame>(games: T[]): SeriesBucket<T>[] {
  if (!games.length) return []

  const sorted = [...games].sort(compareSeriesGames)
  const [teamA, teamB] = canonicalSeriesTeams(sorted[0]!.winner, sorted[0]!.loser)
  const orderedGames = clusterByDate(sorted).flatMap((cluster) =>
    orderSeriesGames(cluster, teamA, teamB),
  )

  const buckets: SeriesBucket<T>[] = []
  let current: T[] = []
  let sessionIndex = 0

  for (let i = 0; i < orderedGames.length; i++) {
    const g = orderedGames[i]!

    if (!current.length) {
      current = [g]
      continue
    }

    if (shouldBreakSeries(current, g, teamA, teamB)) {
      buckets.push({ teamA, teamB, games: current, sessionIndex })
      sessionIndex += 1
      current = [g]
    } else {
      current.push(g)
    }
  }

  if (current.length) {
    buckets.push({ teamA, teamB, games: current, sessionIndex })
  }

  return buckets
}

/**
 * Split chronologically sorted individual games into separate Bo3/Bo5 series.
 * Groups by team pair first so interleaved games from other matchups never break a series.
 */
export function groupGamesIntoSeries<T extends ChronologicalGame>(
  games: T[],
): SeriesBucket<T>[] {
  if (!games.length) return []

  const sorted = [...games].sort(compareSeriesGames)
  const byPair = new Map<string, T[]>()

  for (const g of sorted) {
    const [a, b] = canonicalSeriesTeams(g.winner, g.loser)
    const key = seriesKey(a, b)
    const list = byPair.get(key) ?? []
    list.push(g)
    byPair.set(key, list)
  }

  const buckets: SeriesBucket<T>[] = []
  for (const pairGames of byPair.values()) {
    buckets.push(...splitPairGamesIntoSeries(pairGames))
  }

  return buckets.sort((a, b) => {
    const aFirst = a.games[0]!
    const bFirst = b.games[0]!
    return compareSeriesGames(aFirst, bFirst)
  })
}

File: scripts/ml/series_grouping.py
"""
Group individual OE games into Bo3/Bo5 series — Python port of
src/lib/seriesGrouping.ts, kept behaviorally identical so the ML feature mart's
series grain matches what nucky.gg's recap/series pages consider one series.

Algorithm: pair games by (winner, loser) canonical team pair, cluster by date,
search game-id-ordinal permutations for a valid Bo3/Bo5 win progression, then
split into series buckets whenever the day gap exceeds SERIES_GAP_DAYS or the
series already reached a valid terminal score.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from datetime import date
from itertools import permutations

SERIES_GAP_DAYS = 5
_MAX_PERM_GAMES = 7  # matches the TS guard against permutation blowup

_TAIL_NUM_RE = re.compile(r"[_-](\d+)$")
_EMBEDDED_GAME_RE = re.compile(r"game[_-]?(\d+)", re.IGNORECASE)


@dataclass
class ChronoGame:
    id: str
    game_date: date
    winner: str
    loser: str
    payload: dict = field(default_factory=dict)


@dataclass
class SeriesBucket:
    team_a: str
    team_b: str
    games: list[ChronoGame]
    session_index: int


def series_key(a: str, b: str) -> str:
    return "|".join(sorted([a, b]))


def canonical_series_teams(a: str, b: str) -> tuple[str, str]:
    return tuple(sorted([a, b]))  # type: ignore[return-value]


def days_between(a: date, b: date) -> float:
    return abs((a - b).days)


def game_ordinal_from_id(gid: str) -> int:
    tail = _TAIL_NUM_RE.search(gid)
    if tail:
        return int(tail.group(1))
    embedded = _EMBEDDED_GAME_RE.search(gid)
    if embedded:
        return int(embedded.group(1))
    return 0


def compare_series_games(a: ChronoGame, b: ChronoGame) -> int:
    if a.game_date != b.game_date:
        return -1 if a.game_date < b.game_date else 1
    oa, ob = game_ordinal_from_id(a.id), game_ordinal_from_id(b.id)
    if oa != ob:
        return -1 if oa < ob else 1
    return -1 if a.id < b.id else (1 if a.id > b.id else 0)


def _sort_key(g: ChronoGame):
    return (g.game_date, game_ordinal_from_id(g.id), g.id)


def count_series_wins(games: list[ChronoGame], team: str) -> int:
    return sum(1 for g in games if g.winner == team)


def is_valid_series_score(w_a: int, w_b: int) -> bool:
    hi, lo = max(w_a, w_b), min(w_a, w_b)
    if hi >= 3:
        return hi == 3 and lo <= 2
    if hi == 2:
        return lo <= 1
    return False


def is_valid_game_order(games: list[ChronoGame], team_a: str, team_b: str) -> bool:
    w_a = w_b = 0
    for i, g in enumerate(games):
        if g.winner == team_a:
            w_a += 1
        elif g.winner == team_b:
            w_b += 1
        else:
            return False
        if w_a == 3 or w_b == 3:
            return i == len(games) - 1
    return is_valid_series_score(w_a, w_b)


def _order_distance(a: list[ChronoGame], b: list[ChronoGame]) -> int:
    index_b = {g.id: i for i, g in enumerate(b)}
    return sum(abs(i - index_b.get(g.id, i)) for i, g in enumerate(a))


def order_series_games(games: list[ChronoGame], team_a: str, team_b: str) -> list[ChronoGame]:
    if len(games) <= 1:
        return games
    by_id = sorted(games, key=_sort_key)
    if is_valid_game_order(by_id, team_a, team_b):
        return by_id
    if len(by_id) > _MAX_PERM_GAMES:
        return by_id
    valid = [list(p) for p in permutations(by_id) if is_valid_game_order(list(p), team_a, team_b)]
    if not valid:
        return by_id
    valid.sort(key=lambda order: _order_distance(order, by_id))
    return valid[0]


def _cluster_by_date(games: list[ChronoGame]) -> list[list[ChronoGame]]:
    sorted_games = sorted(games, key=_sort_key)
    clusters: list[list[ChronoGame]] = []
    current: list[ChronoGame] = []
    for g in sorted_games:
        if not current or current[-1].game_date == g.game_date:
            current.append(g)
        else:
            clusters.append(current)
            current = [g]
    if current:
        clusters.append(current)
    return clusters


def is_series_complete(
    games: list[ChronoGame],
    team_a: str,
    team_b: str,
    next_game: ChronoGame | None = None,
) -> bool:
    w_a = count_series_wins(games, team_a)
    w_b = count_series_wins(games, team_b)
    hi, lo, total = max(w_a, w_b), min(w_a, w_b), w_a + w_b

    if hi == 3:
        return True
    if hi != 2 or not is_valid_series_score(w_a, w_b):
        return False

    bo5_in_progress = (total == 2 and lo == 0) or (total == 3 and lo == 1)
    if not bo5_in_progress:
        return True
    if next_game is None:
        return True
    if days_between(games[-1].game_date, next_game.game_date) > SERIES_GAP_DAYS:
        return True

    day_gap = days_between(games[-1].game_date, next_game.game_date)
    if day_gap >= 1 and total == 3 and lo == 1:
        return True
    return False


def _should_break_series(
    bucket_games: list[ChronoGame], new_game: ChronoGame, team_a: str, team_b: str
) -> bool:
    if not bucket_games:
        return False
    last = bucket_games[-1]
    if days_between(last.game_date, new_game.game_date) > SERIES_GAP_DAYS:
        return True
    return is_series_complete(bucket_games, team_a, team_b, new_game)


def _split_pair_games_into_series(games: list[ChronoGame]) -> list[SeriesBucket]:
    if not games:
        return []
    sorted_games = sorted(games, key=_sort_key)
    team_a, team_b = canonical_series_teams(sorted_games[0].winner, sorted_games[0].loser)

    ordered_games: list[ChronoGame] = []
    for cluster in _cluster_by_date(sorted_games):
        ordered_games.extend(order_series_games(cluster, team_a, team_b))

    buckets: list[SeriesBucket] = []
    current: list[ChronoGame] = []
    session_index = 0
    for g in ordered_games:
        if not current:
            current = [g]
            continue
        if _should_break_series(current, g, team_a, team_b):
            buckets.append(SeriesBucket(team_a, team_b, current, session_index))
            session_index += 1
            current = [g]
        else:
            current.append(g)
    if current:
        buckets.append(SeriesBucket(team_a, team_b, current, session_index))
    return buckets


def group_games_into_series(games: list[ChronoGame]) -> list[SeriesBucket]:
    """Split chronologically sorted individual games into separate Bo3/Bo5 series."""
    if not games:
        return []
    by_pair: dict[str, list[ChronoGame]] = {}
    for g in sorted(games, key=_sort_key):
        a, b = canonical_series_teams(g.winner, g.loser)
        by_pair.setdefault(series_key(a, b), []).append(g)

    buckets: list[SeriesBucket] = []
    for pair_games in by_pair.values():
        buckets.extend(_split_pair_games_into_series(pair_games))

    buckets.sort(key=lambda bkt: _sort_key(bkt.games[0]))
    return buckets
